pilot/pkg/bootstrap/server.go

```diff
@@ -591,29 +576,64 @@ func (s *Server) createK8sServiceControllers(serviceControllers *aggregate.Contr
 	return
 }
 
-// initMultiClusterController initializes multi cluster controller
-// currently implemented only for kubernetes registries
-func (s *Server) initMultiClusterController(args *PilotArgs) error {
-	if hasKubeRegistry(args) {
-		s.addStartFunc(func(stop <-chan struct{}) error {
-			secretController := clusterregistry.NewController(s.kubeClient,
-				args.Config.ClusterRegistriesNamespace,
-				s.clusterStore,
-				s.ServiceController,
-				s.EnvoyXdsServer,
-				args.Config.ControllerOptions.ResyncPeriod,
-				args.Config.ControllerOptions.WatchedNamespace,
-				args.Config.ControllerOptions.DomainSuffix)
-
-			// Start secret controller which watches for runtime secret Object changes and adds secrets dynamically
-			go secretController.Run(stop)
+// initMultiClusterSecretController initializes multi cluster secret controller
+// to watch for remote cluster additions or deletions.
+// Currently implemented only for kubernetes registries
+func (s *Server) initMultiClusterSecretController(args *PilotArgs) (err error) {
+	if checkForKubernetes(args.Service.Registries) {
+
+		err = secretcontroller.StartSecretController(s.kubeClient, s.addMemberCluster, s.deleteMemberCluster, args.Config.ClusterRegistriesNamespace)
+		if err != nil {
+			return fmt.Errorf("could not start secret controller: %v", err)
+		}
 
-			return nil
-		})
 	}
 	return nil
 }
 
+//addMemberCluster is passed to the secret controller as a callback to be called
+//when a remote cluster is added
+func (s *Server) addMemberCluster(clientset kubernetes.Interface, clusterID string) error {
+
+	stopCh := make(chan struct{})
+	kubectl := kube.NewController(clientset, kube.ControllerOptions{
+		WatchedNamespace: s.WatchedNamespace,
+		ResyncPeriod:     s.ResyncPeriod,
+		DomainSuffix:     s.DomainSuffix,
+	})
+	s.kubeRegistry[clusterID] = kubectl
```

Comment:
Do not see a lock ?

Reply:
Adding/removing a cluster is done in a single thread (the secret controller thread), so no lock is needed.